consensus/state_processing/src/per_block_processing/eip4844/eip4844.rs

```diff
@@ -0,0 +1,119 @@
+    block_body: BeaconBlockBodyRef<T, Payload>,
+) -> Result<(), BlockProcessingError> {
+    if let (Ok(payload), Ok(kzg_commitments)) = (
+        block_body.execution_payload(),
+        block_body.blob_kzg_commitments(),
+    ) {
+        if let Some(transactions) = payload.transactions() {
+            if !verify_kzg_commitments_against_transactions::<T>(transactions, kzg_commitments)? {
+                return Err(BlockProcessingError::BlobVersionHashMismatch);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+pub fn verify_kzg_commitments_against_transactions<T: EthSpec>(
+    transactions: &Transactions<T>,
+    kzg_commitments: &VariableList<KzgCommitment, T::MaxBlobsPerBlock>,
+) -> Result<bool, BlockProcessingError> {
+    let nested_iter = transactions
+        .into_iter()
+        .filter(|tx| {
+            tx.get(0)
+                .map(|tx_type| *tx_type == BLOB_TX_TYPE)
+                .unwrap_or(false)
+        })
+        .map(|tx| tx_peek_blob_versioned_hashes::<T>(tx));
+
+    itertools::process_results(nested_iter, |mut iter| {
+        let zipped_iter = iter
+            .flatten()
+            // Need to use `itertools::zip_longest` here because just zipping hides if one iter is shorter
+            // and `itertools::zip_eq` panics.
+            .zip_longest(kzg_commitments.into_iter())
+            .enumerate()
+            .map(|(index, next)| match next {
+                EitherOrBoth::Both(hash, commitment) => Ok((hash?, commitment)),
+                // The number of versioned hashes from the blob transactions exceeds the number of
+                // commitments in the block.
+                EitherOrBoth::Left(_) => Err(BlockProcessingError::BlobNumCommitmentsMismatch {
+                    commitments_processed_in_block: index,
+                    commitments_processed_in_transactions: index + 1,
+                }),
+                // The number of commitments in the block exceeds the number of versioned hashes
+                // in the blob transactions.
+                EitherOrBoth::Right(_) => Err(BlockProcessingError::BlobNumCommitmentsMismatch {
+                    commitments_processed_in_block: index + 1,
+                    commitments_processed_in_transactions: index,
+                }),
+            });
+
+        itertools::process_results(zipped_iter, |mut iter| {
+            iter.all(|(tx_versioned_hash, commitment)| {
+                tx_versioned_hash == kzg_commitment_to_versioned_hash(commitment)
+            })
+        })
+    })?
+}
+
+/// Only transactions of type `BLOB_TX_TYPE` should be passed into this function.
+fn tx_peek_blob_versioned_hashes<T: EthSpec>(
+    opaque_tx: &Transaction<T::MaxBytesPerTransaction>,
+) -> Result<
+    impl IntoIterator<Item = Result<VersionedHash, BlockProcessingError>> + '_,
+    BlockProcessingError,
+> {
+    let tx_len = opaque_tx.len();
+    let message_offset = 1 + u32::from_ssz_bytes(opaque_tx.get(1..5).ok_or(
```

Comment:
my spidey sense are tingling for all this arithmetic not using `safe_arith`, but we can patch that up later

Reply:
good call, added it here: 7f54ab8a5b9d9d96de07b7f2ef63fb8659670ba1